src/tool/subcommands/api_cmd/api_run_tests.rs

```diff
@@ -0,0 +1,463 @@
+
+        let last_block = client.call(EthBlockNumber::request(())?).await?;
+
+        let filter_spec = EthFilterSpec {
+            from_block: Some(format!("0x{:x}", last_block.0 - BLOCK_RANGE)),
+            to_block: Some(last_block.to_hex_string()),
+            ..Default::default()
+        };
+
+        let filter_id = client.call(EthNewFilter::request((filter_spec,))?).await?;
+
+        let removed = client
+            .call(EthUninstallFilter::request((filter_id.clone(),))?)
+            .await?;
+        anyhow::ensure!(removed);
+
+        let removed = client
+            .call(EthUninstallFilter::request((filter_id,))?)
+            .await?;
+        anyhow::ensure!(!removed);
+
+        Ok(())
+    })
+}
+
+fn create_eth_new_filter_limit_test(count: usize) -> RpcTestScenario {
+    RpcTestScenario::basic(move |client| async move {
+        const BLOCK_RANGE: u64 = 200;
+
+        let last_block = client.call(EthBlockNumber::request(())?).await?;
+
+        let filter_spec = EthFilterSpec {
+            from_block: Some(format!("0x{:x}", last_block.0 - BLOCK_RANGE)),
+            to_block: Some(last_block.to_hex_string()),
+            ..Default::default()
+        };
+
+        let mut ids = vec![];
+
+        for _ in 0..count {
+            let result = client
+                .call(EthNewFilter::request((filter_spec.clone(),))?)
+                .await;
+
+            match result {
+                Ok(filter_id) => ids.push(filter_id),
+                Err(e) => {
+                    // Cleanup any filters created so far to leave a clean state
+                    for id in ids {
+                        let removed = client.call(EthUninstallFilter::request((id,))?).await?;
+                        anyhow::ensure!(removed);
+                    }
+                    anyhow::bail!(e)
+                }
+            }
+        }
+
+        for id in ids {
+            let removed = client.call(EthUninstallFilter::request((id,))?).await?;
+            anyhow::ensure!(removed);
+        }
+
+        Ok(())
+    })
+}
+
+fn eth_new_block_filter() -> RpcTestScenario {
+    RpcTestScenario::basic(move |client| async move {
+        let filter_id = client.call(EthNewBlockFilter::request(())?).await?;
+
+        let filter_result = client
+            .call(EthGetFilterChanges::request((filter_id.clone(),))?)
+            .await?;
+
+        let result = if let EthFilterResult::Hashes(prev_hashes) = filter_result {
+            let verify_hashes = async |hashes: &[EthHash]| {
+                for hash in hashes {
+                    let _block = client
+                        .call(EthGetBlockByHash::request((hash.clone(), false))?)
+                        .await?;
+                }
+                Ok::<(), crate::rpc::ClientError>(())
+            };
+            verify_hashes(&prev_hashes).await?;
+
+            // Wait till the next block arrive
+            sleep(Duration::from_secs(EPOCH_DURATION_SECONDS as u64)).await;
```

Comment:
Again, I'd avoid sleeps. What if Forest hiccups for some reason? Or maybe there's a re-org? Can we do it in more deterministic (and potentially faster) way?

Reply:
Yes, we can wait for the message to be included using StateWaitMsg, or subscribe to a notification to wait for the next block.